Add GLOBAL_DEF_INTERNAL to hide specific settings

KoBeWi opened the pull request:
I discovered this line:
https://github.com/godotengine/godot/blob/4d2ecde3e22e2503715b0cec612dd8b767dce77a/core/config/project_settings.cpp#L347
which is probably the best way to filter out properties from Project Settings. For unknown reason though, it was checking `/something` instead of `something/` for some settings (which also caused autoloads to not be hidden previously, which was then patched somewhere else). There was also `export/` filter, which was probably unused since `export_presets.cfg` was added. And then there is `remap/` and `locale/` filter. It wasn't working, but seems like locale group has some settings that should be visible, so I removed this and only hidden 2 settings from that group (but maybe there is more settings that could be hidden).

Resolves #58456
Alternative to #61450

KoBeWi committed:
Add GLOBAL_DEF_INTERNAL to hide specific settings

KoBeWi commented on core/config/project_settings.cpp:
```diff
@@ -344,7 +349,7 @@ void ProjectSettings::_get_property_list(List<PropertyInfo> *p_list) const {
 		vc.order = v->order;
 		vc.type = v->variant.get_type();
-		if (vc.name.begins_with("input/") || vc.name.begins_with("import/") || vc.name.begins_with("export/") || vc.name.begins_with("/remap") || vc.name.begins_with("/locale") || vc.name.begins_with("/autoload")) {
+		if (v->internal || vc.name.begins_with("input/") || vc.name.begins_with("import/") || vc.name.begins_with("autoload/") || vc.name.begins_with("editor_plugins/") || vc.name.begins_with("shader_globals/")) {
```
Internal properties are for use only if there are single/few properties in the group that you want to hide. Meantime e.g. `input/` filters out whole input map; there is no single property that you'd want to edit outside the dedicated tab.

We could maybe add a way to define groups that should be hidden, but this code is in project settings, so it's probably ok like this.

akien-mga replied:
Well, it feels weird to have a system to mark properties internal that's only used for a couple properties but not for the many which are actually being explicitly skipped.

Why can't we call `set_as_internal` when registering input mappings or autoloads?

KoBeWi replied:
I could do that. It's just less error prone to do it for the whole group if you know that none of the properties need to be visible.

akien-mga replied:
My assumption is that all those "groups" of properties are actually defined in `for` loops and thus marking them internal is simpler than filtering them later on:
```
diff --git a/core/config/project_settings.cpp b/core/config/project_settings.cpp
index ff5ff83bf8..b91b2b64da 100644
--- a/core/config/project_settings.cpp
+++ b/core/config/project_settings.cpp
@@ -1151,7 +1151,7 @@ void ProjectSettings::_add_builtin_input_map() {
 			action["events"] = events;
 
 			String action_name = "input/" + E.key;
-			GLOBAL_DEF(action_name, action);
+			GLOBAL_DEF_INTERNAL(action_name, action);
 			input_presets.push_back(action_name);
 		}
 	}
```

That being said, for default input actions for example, we do want them to appear in the docs, so it can be different from other cases.

KoBeWi replied:
This loop unfortunately doesn't include user-defined actions. I made these internal, but the filter will still be there.
Same is with other properties filtered here. They are dynamic and user-created and I don't really see other way to hide them, as they are "created" with `set()`.

akien-mga commented on core/config/project_settings.cpp:
```diff
@@ -1235,6 +1241,11 @@ ProjectSettings::ProjectSettings() {
 	GLOBAL_DEF("compression/formats/gzip/compression_level", Compression::gzip_level);
 	custom_prop_info["compression/formats/gzip/compression_level"] = PropertyInfo(Variant::INT, "compression/formats/gzip/compression_level", PROPERTY_HINT_RANGE, "-1,9,1");
+
+	// These properties will not show up in the dialog nor in the documentation. If you want to exclude whole groups, see _get_property_list() method.
+	GLOBAL_DEF_INTERNAL("application/config/features", PackedStringArray());
+	GLOBAL_DEF_INTERNAL("internationalization/locale/translation_remaps", PackedStringArray());
+	GLOBAL_DEF_INTERNAL("internationalization/locale/translations", PackedStringArray());
```
They do show up in the documentation apparently. https://github.com/godotengine/godot/runs/6756182637?check_suite_focus=true

KoBeWi replied:
I noticed that. I'll try other flags.

KoBeWi commented:
Pushed some changes. I made the built-in input actions into internal properties and added docs for the internal properties added in this PR (it's ok to document them I think).

As for the group filter line, it has to stay, because it handles dynamically created properties. I modified it further as `import/` isn't actually used - changed it to `importer_defaults/`.

akien-mga commented on doc/classes/ProjectSettings.xml:
```diff
@@ -811,6 +814,12 @@
 			If non-empty, this locale will be used when running the project from the editor.
 		</member>
+		<member name="internationalization/locale/translation_remaps" type="PackedStringArray" setter="" getter="">
+			Locale-dependent resource remaps. Edit them in the "Localization" tab of Project Settings editor. 
```
Fails CI, trailing whitespace?

KoBeWi replied:
Indeed, I wish this didn't display in such a confusing way :/
![image](https://user-images.githubusercontent.com/2223172/172375942-ca30b1fc-aa0c-4e36-ab89-05a249d885a9.png)

akien-mga reviewed (approved):
Seems fine for the time being, though in discussion @YuriSizov and I agreed that it feels like we could go further in dehardcoding the prefixes in `_get_property_list`.

One option could be to add something like `ProjectSettings::get_singleton()->set_prefix_to_hide(String path, bool hide)` so that the code that adds such settings can be responsible for hiding them too (which makes it available to plugins).

akien-mga commented:
Thanks!
